ofrak_core/ofrak/gui/script_builder.py

```diff
@@ -0,0 +1,348 @@
+from ofrak.core.filesystem import FilesystemEntry
+from ofrak.model.resource_model import Data
+from ofrak.service.resource_service_i import ResourceAttributeValueFilter, ResourceFilter
+from black import format_str, FileMode
+
+from ofrak.resource import Resource
+
+
+class SelectableAttributesError(Exception):
+    """
+    Prompt the user for an attribute to select with
+    """
+
+
+class ActionType(IntEnum):
+    UNPACK = 0
+    MOD = 1
+    PACK = 2
+    UNDEF = 3
+
+
+@dataclass(frozen=True)
+class ScriptAction:
+    """
+    Encapsulates the structure of a single action within the script, which consists of the string
+    representation of the code for that action and the action's type.
+    """
+
+    action_type: ActionType
+    action: str
+
+
+@dataclass
+class ScriptSession:
+    """
+    A script, consisting of an ordered sequence of script actions and a mapping between resources
+    and their autogenerated variable names.
+    """
+
+    actions: List[ScriptAction] = field(default_factory=list)
+    resource_variable_names: Dict[bytes, str] = field(default_factory=dict)
+    boilerplate_header: str = r"""
+    from ofrak import *
+    from ofrak.core import *
+
+    async def main(ofrak_context: OFRAKContext):"""
+    # TODO: Replace with backend in use by OFRAK instance used to create the script.
+    boilerplate_footer: str = r"""
+    if __name__ == "__main__":
+        if False:
+            import ofrak_angr
+            import ofrak_capstone
+            ofrak.discover(ofrak_capstone)
+            ofrak.discover(ofrak_angr)
+
+        if False:
+            import ofrak_binary_ninja
+            import ofrak_capstone
+            ofrak.discover(ofrak_capstone)
+            ofrak.discover(ofrak_binary_ninja)
+
+        if False:
+            import ofrak_ghidra
+            ofrak.discover(ofrak_ghidra)
+
+        ofrak = OFRAK()
+        ofrak.run(main)
+    """
+
+
+class ScriptBuilder:
+    """
+    Builds and maintains runnable OFRAK scripts as sequences of actions, with each script tied to
+    a session.
+    """
+
+    def __init__(self):
+        self.root_cache: Dict[bytes, Resource] = {}
+        self.script_sessions: Dict[bytes, ScriptSession] = {}
+        self.selectable_indexes: List[ResourceIndexedAttribute] = [
+            FilesystemEntry.Name,
+            Data.Offset,
+        ]
+
+    async def add_action(
+        self,
+        resource: Resource,
+        action: str,
+        action_type: ActionType,
+    ) -> None:
+        """
+        Adds an action to the script session to which the selected resource belongs. An action is
+        a string representing the code that is being run on the resource based on an action that
+        has occurred in the GUI.
+
+        :param resource: Resource upon which the action is being taken.
+        :param action: A string describing the code being run based on a GUI action.
+        :param action_type: An instance of `ActionType` categorizing the action.
```

Comment:
I really like the docstrings! They're quite helpful. 

Please make sure they match with the [OFRAK contributor guide style conventions](https://ofrak.com/docs/contributor-guide/getting-started.html#functions-and-methods)!

Reply:
I've added `raises` to `add_variable`, removed the periods from non-general descriptions, and added backticks are what I consider to be "named code artifacts" (slightly unsure of what 'artifacts' means in this context, since nothing is being left behind).